text/3268-repr-scalable.md

```diff
@@ -0,0 +1,159 @@
+accepts an integer to determine the number of elements per granule. See the definitions in
+[the reference-level explanation](#reference-level-explanation) for more information.
+
+e.g. for a scalable vector f32 type the following could be its representation:
+
+```rust
+#[repr(simd, scalable(4))]
+#[derive(Clone, Copy)]
+pub struct svfloat32_t {
+    _ty: [f32; 0],
+}
+```
+`_ty` is purely a type marker, used to get the element type for the LLVM backend.
+
+
+This new class of type has some restrictions on it that a normal ZST wouldn't have, and some of the restrictions that a ZST
+has do not apply to this new type.
+As this type does have a run-time size it can be stored to memory, this is required for spilling to the stack for instance.
+This new class of type can't be stored in a structure or a compound type, as the layout of that wouldn't be known at compile
+time.
+
+
+A simple example that an end user would be able to write for summing of two arrays using functions from the ACLE
+for SVE is shown below:
+```rust
+unsafe {
+    let step = svcntw() as usize;
+    for i in (0..SIZE).step_by(step) {
+        let a = data_a.as_ptr().add(i);
+        let b = data_b.as_ptr().add(i);
+        let c = &mut data_c as *mut f32;
+        let c = c.add(i);
+
+        let pred = svwhilelt_b32(i as _, SIZE as _);
+        let sva = svld1_f32(pred, a);
+        let svb = svld1_f32(pred, b);
+        let svc = svadd_f32_m(pred, sva, svb);
+
+        svst1_f32(svc, pred, c);
+    }
+}
+```
+As can be seen by that example the end user wouldn't necessarily interact directly with the changes that are
+proposed by this RFC, but might use types and functions that depend on them.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This will focus on LLVM. No investigation has been done into the alternative codegen back ends. At the time of
+writing I believe cranelift doesn't support scalable vectors ([current proposal](https://github.com/bytecodealliance/rfcs/pull/19)),
+and the GCC backend is not mature enough to be thinking about this.
+
+Most of the complexity of SVE will be handled by LLVM and the `vscale` modifier that is applied to vector types. Therefore
+changes for this should be fairly minimal for Rust. From the LLVM side this is as simple as calling `LLVMScalableVectorType`
+rather than `LLVMVectorType`.
+
+For a Scalable Vector Type LLVM takes the form `<vscale x elements x type>`.
+* `elements` multiplied by sizeof(`type`) gives the smallest allowed register size and the increment size.
+* `vscale` is a run time constant that is used to determine the actual vector register size.
+
+For example, with Arm SVE the scalable vector register (Z register) size has to be a multiple of 128 bits, therefore for `f32`, `elements` would
+always be four. At run time `vscale` could be 1, 2, 3, through to 16 which would give register sizes of 128, 256, 384 to 2048.
```

Comment:
Is 0 a valid vscale for any architecture?

Reply:
no afaict. you can set VL to 0, but vscale is instead basically VLMAX. vscale must be positive on SVE[2] and RVV. idk about SX Aurora but I'd expect vscale is always 64.

for [SVP64 (SimpleV)](https://libre-soc.org/openpower/sv/) instead of the cpu selecting VLMAX, the programmer selects it, so it'd work better to use fixed-size vectors (`Simd<MaybeUninit<T>, VLMAX>` with the first `len`/`VL` elements used, so kinda like `ArrayVec`) rather than scalable vectors (though we're thinking of also supporting compiler-selected VLMAX for compatibility with RVV/SVE, this would use scalable vector types). VLMAX must be in `1..=64` (or `1..=128` or more for future extensions). the instruction for selecting `VLMAX` ([`setvl`](https://libre-soc.org/openpower/sv/setvl/)) doesn't even have a way to encode VLMAX=0, it encodes VLMAX=1 using all zero bits in the appropriate immediate field.